### DISCUSSION TITLE
is the code stabilized?

### DISCUSSION THREAD

**Discussion by manson** [kamalshkeir/kago#5] [Q&A] [answered]
I mean may the code change brining API breaking changes that influence the existing apps based on kago?

**Comment by kamalshkeir** [comment, marked as answer]
Hello, no break changes will be introduced

**Reply by manson** [reply]
From one point of view this is good - we can use it with no worries on future breakings, from the second point of view - does this mean that no further development happen? As far as I remember you was about to attach NoSQL databases (mongo), probably there could be other improvements?

**Reply by kamalshkeir** [reply]
Does this mean that no further development happen?
The answer is no.
I will add Mongo, and should not break any existing code, but i can say, it will be the hardest feature to add.
Other features will be added of course, like:
- a configstore package that do pretty much same thing as envloader, but for json, yaml,toml,... , loaded to struct directly
- shell can be improved, i will make adding a custom commands to shell possible
- tests should be added
- godocs should be improved , i can put more examples

I am using kago in production for my portfolio/blog, so i will try to keep my live website up to date with the last version of kago, and make it like my demo/test production website, i monitor with grafana pretty much everything and i already tested many security vulnerabilities on it for csrf and cors.

Finaly, i was wondering , what do you think about making a reddit post , challenging gophers to break into my website https://kamalshkeir.dev, i donno, i think it can be helpful to find very fast all security issues, or it's a bad idea, lol :smile:

**Reply by manson** [reply]
Very nice.
As for the security issues on you site - I think this will be useful. Personally I'm not like that "hacker" that can do such a tests but who knows, maybe some people from the reddit may consider this as interesting. 
As for the Kago itself - I find it very interesting and useful. Currently I don't have new projects to use this but definitely the next project (large or small) will use Kago as base. The only two disappointing thinks personally for me is absence the mongo implementation (my base database I use) and not quite interesting admin panel it has right now. Yes, I remember that I can write my own and more or less I can use current admin panel for quick fixes but... It looks not like integral part of any application and a very  simple. I would rather prefer to think of my own app application logic instead of spending many time to implement own admin panel than needs to be in any middle or higher app or site and that should look nice especially for a customer.  But ok, I understand that that is not the main purpose of this app. Anyway, it is a great job you have fulfilled!

**Reply by kamalshkeir** [reply]
Hello my friend,
Glad to hear from you again.
As for security vulnerabilities : https://www.reddit.com/r/golang/comments/xdzuo4/challenge_for_hackers_to_test_security_of_a/ , hope someone help with that.
As for mongo, i was thinking about how to implement it, if i do it with ORM and sqlbuilder , many changes should be done, and it can cause problem with migrations, so the best way i found is to make a separated package for nosql databases, other than ORM, like NORM for example, what do you think ?
You really have problem with this admin panel :laughing: ,  but i don't understand why ?, the admin panel is like django admin, need to be functional for developers, but not to be used for customers like a wordpress dashboard for instance, 
So if it is a matter of style, it is very hard to find a dashboard design that fit every developer or every customer style and needs, but if you have ideas about a new design that fit better or new functionalities to add to the admin why not, i'm sure we can achieve this framework without any disappointing thinks :smile:

**Reply by manson** [reply]
"As for mongo...it can cause problem with migrations" - why? When using mongo you need no migrations at all! And I think it is even easier to implement it usage than for relational sql. If use not official mongo driver that is hard, but something like https://golangexample.com/qmgo-the-go-driver-for-mongodb-its-based-on-official-mongo-go-driver-but-easier-to-use-like-mgo/ I think it is very close to your orm implementation
As for the admin panel, it is always good to have something like pocketbase has or other frameworks to be named"professional" :-)

" the admin panel is like django admin, need to be functional for developers, but not to be used for customers like a wordpress dashboard for instance," - I don't use django and even never touched it so don't know what your'e talking about. :-) But I have a usefull extendable admin panel for my own old site (nodejs-based) that has a lot of functionality to administrate it. And I cannot imagine how I would use your panel for the same purpose. If I need something like this I would rather will have to implement it here from the ground. And it is a time to spend. For example here is screenshot of my admin, it is in russian but it does not matter here. This framework had initial admin functionality that I could extend when needed. In your case  I'll have to cook something up of my own
![image](https://user-images.githubusercontent.com/176285/190193841-3e96cfc9-784a-4d94-a7c3-e11b1b25b7f8.png)

As I said before in previous coversation, I think go-admin would go perfect here: https://github.com/GoAdminGroup/go-admin

....

**Reply by kamalshkeir** [reply]
As for mongo db, i spent all the night yesterday trying to implement it using std driver not Qmgo, it was a nightmare, changes should be done everywhere, until i realise that qmgo is an orm like, so you can use it beside sqlite, but after many hours i was facing a logical problem , how can i handle nosql not consistent data in a tables ?, columns may change from one row to another in the same table because there is no schema, so i take a look to pocketbase , they use sql only, even go-admin it only support sql, so i think that's why it's not convenient to use schemaless database for the admin except if i change admin tables to like json formated rows.
So after all i think this is the reason why schemaless databases use json format in there explorer

As for the admin, more features can and will be added in the futur, like search and filters, permissions, changing passwords, the only limit is our imagination really , even the style, and as i said before it is now fully editable, you can modify whathever you don't like, push to a repos, and use your repos for all your projects, or create one time your perfect admin dash that you can , and others can used by using the same mechanism

As for go-admin, i'm sorry but i really don't like it, my dashboard use pure js, it's PWA, CRUD, the alert window written using custom html elements, infinite scroll using intersectionObserver, notifications, all tools are easy to edit under assets/static/tools, because i don't like to use external libraries, look to what go-admin use, do you like to have all this jquery in your binary ? :
Dependencies
[Datetimepicker](http://eonasdan.github.io/bootstrap-datetimepicker/)
[font-awesome](http://fontawesome.io/)
[bootstrap-fileinput](https://github.com/kartik-v/bootstrap-fileinput)
[jquery-pjax](https://github.com/defunkt/jquery-pjax)
[Nestable](http://dbushell.github.io/Nestable/)
[toastr](http://codeseven.github.io/toastr/)
[bootstrap-number-input](https://github.com/wpic/bootstrap-number-input)
[fontawesome-iconpicker](https://github.com/itsjavi/fontawesome-iconpicker)

**Reply by manson** [reply]
"how can i handle nosql not consistent data in a tables ?, columns may change from one row to another in the same table because there is no schema" - this doesn't matter at all. Your current structure is your classes being used to store data. If no columns in db but exists in structure when querying - it just left empty. If no column on saving - it just created for this data. No need to control the data and structures consistency. This is all the developer responsibility. It may sound awful but it never gave any problems before just because this is something developer knows and he knows how to deal with this. Of course  this depends on application type being used, for example it is not good idea to use NoSQL for financial apps or where transactions are mandatory. But in most sites (and many business apps) it is a good choice and comfortable. I've used both database types for years and has no any problems in both technologies usage.

"As for go-admin, i'm sorry but i really don't like it" - no problem. But if you don't like any external dependencies, you have to create admin panel from the ground up. Not sure this is a good idea unless this is something you really want to do on pure js. After all there is a lot of separate admins on a market - from pure html pages with css to react, vue, svelte or others more robust panels but many of them uses external libs. Many are free, some are payed....

" or create one time your perfect admin dash that you can" - sure I can :-) And I did this many times before as from the pure js-ed till vue, go-admin and others. And I'm tired of all this and this is not I'm happy to do once again :-)  Well... maybe some people appear who would do this for all others and for me of cause :-) Anyway, this is not something to prevent from using Kargo :-)
